cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -385,41 +391,27 @@ def ceph_storage(ceph_admin):
     set_state('ceph-storage.configured')
 
 
-def launch_dns():
-    '''Create the "kube-system" namespace, the kubedns resource controller, and
-    the kubedns service. '''
-    message = 'Rendering the Kubernetes DNS files.'
-    hookenv.log(message)
-    hookenv.status_set('maintenance', message)
-    # Render the DNS files with the cider information.
-    render_files()
-    # Run a command to check if the apiserver is responding.
-    return_code = call(split('kubectl cluster-info'))
-    if return_code != 0:
-        hookenv.log('kubectl command failed, waiting for apiserver to start.')
-        remove_state('kube-dns.available')
-        # Return without setting kube-dns.available so this method will retry.
-        return
-    # Check for the "kube-system" namespace.
-    return_code = call(split('kubectl get namespace kube-system'))
-    if return_code != 0:
-        # Create the kube-system namespace that is used by the kubedns files.
-        check_call(split('kubectl create namespace kube-system'))
-    addon_dir = '/etc/kubernetes/addons'
-    # Check for the kubedns replication controller.
-    get = 'kubectl get -f {0}/kubedns-rc.yaml'.format(addon_dir)
-    return_code = call(split(get))
-    if return_code != 0:
-        # Create the kubedns replication controller from the rendered file.
-        create = 'kubectl create -f {0}/kubedns-rc.yaml'.format(addon_dir)
-        check_call(split(create))
-    # Check for the kubedns service.
-    get = 'kubectl get -f {0}/kubedns-svc.yaml'.format(addon_dir)
-    return_code = call(split(get))
-    if return_code != 0:
-        # Create the kubedns service from the rendered file.
-        create = 'kubectl create -f {0}/kubedns-svc.yaml'.format(addon_dir)
-        check_call(split(create))
+def create_addon(template, context):
+    '''Create an addon from a template'''
+    target = '/etc/kubernetes/addons/' + template
+    render(template, target, context)
+    cmd = ['kubectl', 'create', '-f', target]
```

Comment:
Good point. I think I'd like to use [kubectl apply](http://kubernetes.io/docs/user-guide/kubectl/kubectl_apply/) here instead:

> Apply a configuration to a resource by filename or stdin. This resource will be created if it doesn’t exist yet.